### PR TITLE
Lopatin Ilya. Lab 2: LLVM IR. Option 1.

### PR DESCRIPTION
**Pull request by ascannel**
This LLVM Pass optimizes floating point arithmetics in LLVM IR by merging `fmul` and `fadd` instructions into `llvm.fmuladd` intrinsic.

### Functionality:
- Merge operations
- Dead code elimination: automatically removes `fmul` and `fadd` instructions when they become unused
- Works with `float` and `double` data types

### LLVM APIs Used:
- `FunctionPass`: Analyzes and modifies functions within LLVM IR.
- `IRBuilder<>`: Constructs intrinsics and manages instruction insertion.
- `BinaryOperator`: Identifies FAdd and FMul operations.
- `Intrinsic::fmuladd`: Generates the intrinsic.
- `PreservedAnalyses`: Tracks whether IR analyses remain valid after transformations.

### Workflow:
1. Plugin Registration: Integrates with LLVM via `PassPlugin` for compatibility with modern passmanager.
2. Basic Block Traversal: Iterates through all basic blocks in a function.
3. Pattern Matching: Detects valid fmul+fadd patterns.
4. Intrinsic Replacement: Uses `IRBuilder` to create `llvm.fmuladd` and replaces the original instructions.
5. Cleanup: Safely removes orphaned fmul and fadd instructions.

### CODE REVIEW

File: llvm/compiler-course/llvm-ir/lopatin_fmul_fadd/fmulFaddMergePass.cpp [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+      }
+
+      for (auto *fadd : toReplace) {
+        llvm::BinaryOperator *fmul = nullptr;
+        llvm::Value *thirdOp = nullptr;
+
+        if (auto *op =
+                llvm::dyn_cast<llvm::BinaryOperator>(fadd->getOperand(0))) {
+          if (op->getOpcode() == llvm::Instruction::FMul) {
+            fmul = op;
```
Comment (m-ly4) [inline llvm/compiler-course/llvm-ir/lopatin_fmul_fadd/fmulFaddMergePass.cpp]: These operands might be collected with fadd in initial search.

File: llvm/test/compiler-course/lopatin_fmul_fadd/testFmulFaddMerge.cpp [hunk 2]
```diff
@@ -0,0 +1,13 @@
+double fmaDouble(double a, double b, double c) {
```
Comment (m-ly4) [inline llvm/test/compiler-course/lopatin_fmul_fadd/testFmulFaddMerge.cpp]: You may add these lines as a comment into main lit test file.

File: llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+define dso_local noundef double @fmaDouble(double %a, double %b, double %c) {
+entry:
+  %mul = fmul double %a, %b
+  %add = fadd double %mul, %c
+  ret double %add
+}
+
+define dso_local noundef float @fmaFloat(float %a, float %b, float %c) {
+entry:
+  %mul = fmul float %a, %b
```
Comment (m-ly4) [inline llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll]: Add the following test:
```
float fmaFloat(float a, float b, float c) {
    float res1 = a * b;
    float res2 = b * c;
    float res3 = res1 + res2;
    return res3;
}
```

**Review by m-ly4 — CHANGES_REQUESTED** [review]

**Comment by m-ly4** [comment]
Please also check clang-tidy check output. There are present warnings that should be applied among false-positive.

File: llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+; CHECK-NOT: fmul double
+; CHECK-NOT: fadd double
+
+define dso_local noundef double @fmaDouble(double %a, double %b, double %c) {
+entry:
+  %mul = fmul double %a, %b
+  %add = fadd double %mul, %c
+  ret double %add
+}
+
```
Comment (aobolensk) [inline llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll:14]: Please, add the test where fma is not fused (negative tests)

File: llvm/compiler-course/llvm-ir/lopatin_fmul_fadd/fmulFaddMergePass.cpp [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+      for (llvm::Instruction &inst : bb) {
+        if (auto *fadd = llvm::dyn_cast<llvm::BinaryOperator>(&inst)) {
+          if (fadd->getOpcode() == llvm::Instruction::FAdd) {
+            if (auto *fmul =
+                    llvm::dyn_cast<llvm::BinaryOperator>(fadd->getOperand(0))) {
+              if (fmul->getOpcode() == llvm::Instruction::FMul) {
+                toReplace.push_back(fadd);
+              }
+            } else if (auto *fmul = llvm::dyn_cast<llvm::BinaryOperator>(
+                           fadd->getOperand(1))) {
```
Comment (aobolensk) [inline llvm/compiler-course/llvm-ir/lopatin_fmul_fadd/fmulFaddMergePass.cpp]: Why `fadd->getOperand(1)` specifically ?

**Review by aobolensk — CHANGES_REQUESTED** [review]

File: llvm/compiler-course/llvm-ir/lopatin_fmul_fadd/fmulFaddMergePass.cpp (hunk 1 above)
Comment (ascannel) [reply llvm/compiler-course/llvm-ir/lopatin_fmul_fadd/fmulFaddMergePass.cpp]: don't really understand, do you mean to check recursive for them? cause now I think it's safer to do it this way. it works only on clear `fmul` + `fadd` pairs, cause, for example, what if we do `fmul` first, then change datatype from double to float, and then do `fadd`? or i don't really understood you. if not, please explain your comment deeper.

File: llvm/compiler-course/llvm-ir/lopatin_fmul_fadd/fmulFaddMergePass.cpp (hunk 5 above)
Comment (ascannel) [reply llvm/compiler-course/llvm-ir/lopatin_fmul_fadd/fmulFaddMergePass.cpp]: cause I search for two types of patterns:
`a + b * c ` or `a * b + c`.
and so this is why I check for `getOperand(0)` or `getOperand(1)`, depending on the position of the `fadd` in pair

File: llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll (hunk 3 above)
Comment (ascannel) [reply llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll]: added, checked it, it showed this:
```
  %mul1 = fmul float %B, %C
  %0 = call float @llvm.fmuladd.f32(float %A, float %B, float %mul1)
  ret float %0
```
check if this looks right to you. for me it is, cause pass actually added intrinsic. or should it be negative test?

File: llvm/test/compiler-course/lopatin_fmul_fadd/testFmulFaddMerge.cpp (hunk 2 above)
Comment (ascannel) [reply llvm/test/compiler-course/lopatin_fmul_fadd/testFmulFaddMerge.cpp]: and delete the `.cpp` file, right? or just add coments to lit test?

Comment (aobolensk) [reply llvm/test/compiler-course/lopatin_fmul_fadd/testFmulFaddMerge.cpp]: Yes, please, delete

File: llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll (hunk 4 above)
Comment (ascannel) [reply llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll:14]: added, i think

File: llvm/compiler-course/llvm-ir/lopatin_fmul_fadd/fmulFaddMergePass.cpp (hunk 1 above)
Comment (m-ly4) [reply llvm/compiler-course/llvm-ir/lopatin_fmul_fadd/fmulFaddMergePass.cpp]: I meant when you are on line 21 or 26 you already knows what is fmul operation and what is position of 3rd operand. Here you do extra work for checking again type of fadd operands.

File: llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll [hunk 6]
```diff
@@ -0,0 +1,67 @@
+; RUN: opt -load-pass-plugin %llvmshlibdir/FmulFaddMergePass_LopatinIlya_FIIT3_LLVM_IR%pluginext \
+; RUN: -passes="FmulFaddMergePass" -S %s | FileCheck %s
+
+; CHECK: call double @llvm.fmuladd.f64
+; CHECK: call float @llvm.fmuladd.f32
```
Comment (m-ly4) [inline llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll]: Please add checks closer to source IR lines.
By the way only those lines in output are checked for CHECK-NOTs that go after both lines met positive CHECKs. In this case if there still left `fmul double` between `call [double|float] fmuladd`, your tests wouldn't detect that fail.

File: llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+; float fmaFloatSecond(float A, float B, float C) {
+;     float Res1 = A * B;
+;     float Res2 = B * C;
+;     float Res3 = Res1 + Res2;
+;     return Res3;
+; }
+
+define dso_local noundef float @fmaFloatSecond(float %A, float %B, float %C) {
```
Comment (m-ly4) [inline llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll]: Where are checks for this case?

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll (hunk 7 above)
Comment (ascannel) [reply llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll]: aren't they the ones on top, this test just under them?

File: llvm/compiler-course/llvm-ir/lopatin_fmul_fadd/fmulFaddMergePass.cpp (hunk 1 above)
Comment (ascannel) [reply llvm/compiler-course/llvm-ir/lopatin_fmul_fadd/fmulFaddMergePass.cpp]: and is it wrong? it's, i think, safer and simpler for me to do it this way. or is it in need of change in your opinion?

File: llvm/test/compiler-course/lopatin_fmul_fadd/testFmulFaddMerge.cpp (hunk 2 above)
Comment (ascannel) [reply llvm/test/compiler-course/lopatin_fmul_fadd/testFmulFaddMerge.cpp]: `.cpp` deleted, comments added

File: llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll (hunk 6 above)
Comment (ascannel) [reply llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll]: probably, but aren't we testing concrete functions with known (as for me) output for them?
so you're requesting to add check-label for each function, did I understand you correctly?

File: llvm/compiler-course/llvm-ir/lopatin_fmul_fadd/fmulFaddMergePass.cpp (hunk 1 above)
Comment (m-ly4) [reply llvm/compiler-course/llvm-ir/lopatin_fmul_fadd/fmulFaddMergePass.cpp]: This works as well, of course. However, it is not so effective and a bit complicates reading of a code. If you doubt that you can improve this code without code quality degradation then ok. I don't insist on this change, just proposed more concise piece of code.

File: llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll (hunk 3 above)
Comment (m-ly4) [reply llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll]: Thank you. This is good resolution of the pass.

File: llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll (hunk 6 above)
Comment (m-ly4) [reply llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll]: Not a check-label specifically, but a separate `CHECK: call (double|float) @llvm.fmuladd.f(64|32)` for each line where it is expected. Because CHECK works not for all lines of output text but only first occurrence. If, for example, you expect 2 lines of fmuladd in output but wrote only one CHECK then it verifies only first occurred line of the pattern and won't check further lines. So, you may miss a moment when after some modification of your pass or other LLVM stuff first or second fmuladd wrongly disappear.

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll (hunk 6 above)
Comment (m-ly4) [reply llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll]: Second moment regarding of writing LIT tests is ease of reading test files by other developers.
It is clearer what line a check verifies if it is closer to it than when all checks are at the beginning of the test.

File: llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll (hunk 7 above)
Comment (ascannel) [reply llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll]: added checks via `check-label` on the top of every test

File: llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll (hunk 6 above)
Comment (ascannel) [reply llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll]: added checks via `check-label` on the top of every test

File: llvm/compiler-course/llvm-ir/lopatin_fmul_fadd/fmulFaddMergePass.cpp (hunk 1 above)
Comment (ascannel) [reply llvm/compiler-course/llvm-ir/lopatin_fmul_fadd/fmulFaddMergePass.cpp]: I think I've got it, it also required me to use my own struct instead of just vector, but now I think it meets your proposition

File: llvm/compiler-course/llvm-ir/lopatin_fmul_fadd/fmulFaddMergePass.cpp (hunk 5 above)
Comment (ascannel) [reply llvm/compiler-course/llvm-ir/lopatin_fmul_fadd/fmulFaddMergePass.cpp]: please check again my pass, so we could come to the agreement

File: llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+; CHECK-NOT: fadd float
+define dso_local noundef float @fmaFloat(float %A, float %B, float %C) {
+entry:
+  %mul = fmul float %A, %B
+  %add = fadd float %mul, %C
+  ret float %add
+}
+
+; CHECK-LABEL: @recursiveTest
+; CHECK: call float @llvm.fmuladd.f32
```
Comment (m-ly4) [inline llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll]: Please add check for all left instructions and var names that used there.

File: llvm/compiler-course/llvm-ir/lopatin_fmul_fadd/fmulFaddMergePass.cpp [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+      // 1st pass, collecting info about replacements
+      for (llvm::Instruction &Inst : BB) {
+        if (auto *FAdd = llvm::dyn_cast<llvm::BinaryOperator>(&Inst)) {
+          if (FAdd->getOpcode() == llvm::Instruction::FAdd) {
+            llvm::Value *Op0 = FAdd->getOperand(0);
+            llvm::Value *Op1 = FAdd->getOperand(1);
+
+            auto CheckOperand =
+                [&](llvm::Value *Operand,
```
Comment (m-ly4) [inline llvm/compiler-course/llvm-ir/lopatin_fmul_fadd/fmulFaddMergePass.cpp]: This lambda can be moved out of for-loops as invariant to iteration arguments.
```suggestion
            auto CheckOperand =
                [](llvm::Value *Operand,
```

File: llvm/compiler-course/llvm-ir/lopatin_fmul_fadd/fmulFaddMergePass.cpp [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+                    llvm::Value *OtherOperand) -> llvm::BinaryOperator * {
+              if (auto *FMul = llvm::dyn_cast<llvm::BinaryOperator>(Operand)) {
+                if (FMul->getOpcode() == llvm::Instruction::FMul) {
+                  // checking that 2nd operand is NOT fmul
```
Comment (m-ly4) [inline llvm/compiler-course/llvm-ir/lopatin_fmul_fadd/fmulFaddMergePass.cpp]: I suppose that in case if both operands are fmuls then any of them could be selected for optimization instead of just ignoring this instruction.

File: llvm/compiler-course/llvm-ir/lopatin_fmul_fadd/fmulFaddMergePass.cpp (hunk 1 above)
Comment (m-ly4) [reply llvm/compiler-course/llvm-ir/lopatin_fmul_fadd/fmulFaddMergePass.cpp]: Yes, it meets, thank you.

File: llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll (hunk 6 above)
Comment (m-ly4) [reply llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll]: Thank you.

File: llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll
Comment (Kuznetsov-Artyom) [inline llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll:1]: Add test please
```cpp
float foo(float a, float b, float c) {
  float t1 = a * b;
  float t2 = t1 + c;
  float t3 = t2 / (t1 + 1);
  return t3;
}
```

File: llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+; CHECK: fadd float %mul1, %mul2
+; CHECK-NOT: call float @llvm.fmuladd.f32
+define float @noFusionTest(float %x, float %y, float %z) {
+entry:
+  %mul1 = fmul float %x, %y
+  %mul2 = fmul float %y, %z
+  %sum = fadd float %mul1, %mul2
+  ret float %sum
+}
+
```
Comment (Kuznetsov-Artyom) [inline llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll]:
```suggestion


```

**Review by Kuznetsov-Artyom — CHANGES_REQUESTED** [review]

**Comment by ascannel** [comment]
as you so aggressively reviewing my PR, I think, there's no chance for me to answer correctly to you on seminar, as it would be just like here, stream of complains...

File: llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll
Comment (ascannel) [reply llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll:1]: added

File: llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll (hunk 8 above)
Comment (ascannel) [reply llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll]: added

File: llvm/compiler-course/llvm-ir/lopatin_fmul_fadd/fmulFaddMergePass.cpp (hunk 9 above)
Comment (ascannel) [reply llvm/compiler-course/llvm-ir/lopatin_fmul_fadd/fmulFaddMergePass.cpp]: moved

File: llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll (hunk 11 above)
Comment (ascannel) [reply llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll]: fixed, at least i think so. there's a huge problem in empty lines, even if I hit "enter" in vscode to add them, there're not visible on github

File: llvm/compiler-course/llvm-ir/lopatin_fmul_fadd/fmulFaddMergePass.cpp (hunk 10 above)
Comment (ascannel) [reply llvm/compiler-course/llvm-ir/lopatin_fmul_fadd/fmulFaddMergePass.cpp]: are we writing an ultimate pass or just a pass with educational purpose in mind? it's the 2nd time you propose to change logics of my pass, and don't you think it's already enough requested changes for one person? if it was rude, I'm really sorry, but we're all humans and we can be tired of doing something and not recieving at least some good response.

File: llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll (hunk 8 above)
Comment (m-ly4) [reply llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll]: Please add operand names of fmuladd also in check.

**Review by Kuznetsov-Artyom — APPROVED** [review]

File: llvm/compiler-course/llvm-ir/lopatin_fmul_fadd/fmulFaddMergePass.cpp (hunk 9 above)
Comment (ascannel) [reply llvm/compiler-course/llvm-ir/lopatin_fmul_fadd/fmulFaddMergePass.cpp]: with moving lambda out of for-loops I'm getting an intrinsic in negative test, so what would we do? keep negative test or moving function instead?

File: llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll (hunk 8 above)
Comment (ascannel) [reply llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll]: added

**Comment by m1likus** [comment]
I'm sorry to interrupt, but looks like this LLVM Pass is going to be contibuted to LLVM Repository, because I don't see another good reasons why exactly this student gets much more comments from teacher(s) than others.

**Comment by m-ly4** [comment]
> as you so aggressively reviewing my PR, I think, there's no chance for me to answer correctly to you on seminar, as it would be just like here, stream of complains...

Sorry for my simple English and a plenty of remarks if they create an impression that I'd like to overwhelm you. My intention was opposite - to give more food for discussion about programming and compiler internals.

**Comment by m-ly4** [comment]
> I'm sorry to interrupt, but looks like this LLVM Pass is going to be contibuted to LLVM Repository, because I don't see another good reasons why exactly this student gets much more comments from teacher(s) than others.

I'd like it were so, but I doubt that community will accept it now. But if we'd spend another couple of months on improvements in here it might become a good candidate)
As for concern that Lopatin Ilya receives more comments than others, then I heard you and agree. I should give other students a chance to show their knowledge and learning drive too.

File: llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll (hunk 8 above)
Comment (m-ly4) [reply llvm/test/compiler-course/lopatin_fmul_fadd/litTestFmulFaddMerge.ll]: Thank you.

File: llvm/compiler-course/llvm-ir/lopatin_fmul_fadd/fmulFaddMergePass.cpp (hunk 10 above)
Comment (m-ly4) [reply llvm/compiler-course/llvm-ir/lopatin_fmul_fadd/fmulFaddMergePass.cpp]: Ok, I see. You've done really a good job in applying all the remarks and improving this code.
Thank you.

**Review by m-ly4 — APPROVED** [review]

**Review by aobolensk — APPROVED** [review]
